tokio-util/src/sync/poll_semaphore.rs

```diff
@@ -83,3 +78,38 @@ impl fmt::Debug for PollSemaphore {
             .finish()
     }
 }
+
+struct PollSemaphoreStream {
+    semaphore: Arc<Semaphore>,
+    next_fut:
+        Pin<Box<dyn Future<Output = Result<OwnedSemaphorePermit, AcquireError>> + Send + Sync>>,
+}
+
+impl PollSemaphoreStream {
+    fn new(semaphore: Arc<Semaphore>) -> Self {
+        let initial_fut = Arc::clone(&semaphore).acquire_owned();
+
+        Self {
+            semaphore,
+            next_fut: Box::pin(initial_fut),
+        }
+    }
+}
+
+impl Stream for PollSemaphoreStream {
+    type Item = OwnedSemaphorePermit;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let next_fut = self.next_fut.as_mut();
+
+        match ready!(next_fut.poll(cx)) {
+            Ok(permit) => {
+                let fut = Arc::clone(&self.semaphore).acquire_owned();
+                drop(mem::replace(&mut self.next_fut, Box::pin(fut)));
```

Comment:
The other solution doesn't allocate a new `Box` for every permit.

Reply:
If that's a deal-breaker then this PR is dead without the the unsafe+channel+thread-local combo from async-stream. It is a real shame to use tokio-util only for the codec types and have syn as an un-featured dependency.

Reply:
You're right, I agree that it was a mistake to merge `PollSemaphore` without a feature flag. I just opened #3464, which is one possible solution.

Reply:
That looks very promising, better you than me to write that unsafe.